### ISSUE TITLE
Warning in chrome:  The MediaStream 'ended' event is deprecated and will be removed in M54, around October 2016

### ISSUE THREAD

**Issue by MichaelJCole** [andyet/SimpleWebRTC#487]
Here's the message:

```
The MediaStream 'ended' event is deprecated and will be removed in M54, around October 2016. See https://www.chromestatus.com/features/5730404371791872 for more details.
```

It looks like it's referencing this code:  https://github.com/andyet/SimpleWebRTC/blob/ecfd2b744753e25ac3189776a7d13183c7d15d74/peer.js#L248 

Chrome Version 52.0.2743.82 (64-bit)

**Comment by gueff** [comment]
And here is described which changes will be required: https://developers.google.com/web/updates/2015/07/mediastream-deprecations

**Comment by tgabi333** [comment]
i would like to create a patch based on this solution: https://github.com/otalk/localmedia/blob/master/localmedia.js#L68

@fippo ?

**Comment by fippo** [comment]
fell through the cracks... go ahead @tgabi333 -- thanks!

**Comment by joshontheweb** [comment]
I'm still getting this error but it is coming from the localmedia module here: https://github.com/andyet/SimpleWebRTC/blob/master/simplewebrtc.bundle.js#L5402

It looks like the latest version of localmedia has fixed this.  Any reason why it hasn't been pulled into the latest simplewebrtc bundle?

**Comment by tgabi333** [comment]
I was not sure wether to upgrade this repository because simplewebrtc inherits from that prototype and exposes the api.

I should fork localmedia repo and release a new 2.x patch version with that fix on the weekend.

**Comment by joshontheweb** [comment]
Many Thanks!

**Comment by janwendt** [comment]
This issue causes that in M54 it is not possible to share the localScreen two times:
After removing the shared screen the MediaStream Object is not removed. Only its property "active" is set from "true" to "false". When I try to reshare the Screen no new MediaStreamTrack is added so it does not work anymore.

**Comment by MichaelJCole** [comment]
@janwendt Hey, I saw your message so decided to try it out.  I was able to share/unshare/share screens on M54 (linux) w/o issue.  I'm running a custom signal server, so that may be where to look for this issue if you're still encountering it.

**Comment by greg-man** [comment]
Same issue here as @janwendt. Latest bundle is still using stream.onended here: https://github.com/andyet/SimpleWebRTC/blob/master/simplewebrtc.bundle.js#L7628. A quick fix would be [watching the MediaStreamTracks instead of MediaStream for the ended event](https://developers.google.com/web/updates/2016/09/chrome-54-deprecations#remove_mediastream_ended_event_and_attribute_and_onended_attribute). Fo example:

```
 stream.getTracks().forEach(function (track) {
                track.addEventListener('ended', function () {
                    var idx = self.localScreens.indexOf(stream);
                    if (idx > -1) {
                        self.localScreens.splice(idx, 1);
                    }
                    self.emit('localScreenStopped', stream);
                });
            });
```

**Comment by fippo** [comment]
is this just an issue with the bundle? I don't recommend using that but updating it is fairly easy

**Comment by greg-man** [comment]
I don't know. Sadly Firefox doesn't seem to like the track.addEventListener patch so you still need the stream.onended routine to have it work in firefox as well. Any one else having this issue? We are using latest simplewebrtc bundle with signalmaster.

**Comment by fippo** [comment]
@pehrsons ^^ you know more about MediaStreamTrack than me

**Comment by Pehrsons** [comment]
Which version are you on @greg-man? I implemented the ended event in firefox 50: https://bugzilla.mozilla.org/show_bug.cgi?id=1208373

**Comment by greg-man** [comment]
Firefox 50.0.2 on Ubuntu 16.04.1 (sorry for the late response btw). Indeed the track ended event isn't fired in Firefox. Just checked that by logging to the console: Chrome 55 does, Firefox 50 doesn't.

**Comment by greg-man** [comment]
Well, we call the stopScreenShare method of the simplewebrtc bundle, located [here](https://github.com/andyet/SimpleWebRTC/blob/master/simplewebrtc.bundle.js#L18218). As you can see each track of the stream is stopped as follows:

`stream.getTracks().forEach(function (track) { track.stop(); });`

Any idea what is going wrong here? Would be nice to have this fixed.

**Comment by Pehrsons** [comment]
stop()ing a track doesn't raise "ended" according to mediacapture-main. Chrome is doing it wrong. You can check track.readyState after stop(), should be "ended".

**Comment by Pehrsons** [comment]
The rationale is that since the application is calling stop() it can keep track of the track's state itself and the event is superfluous. (also stop() is sync but raising "ended" is async - yeah chrome doesn't follow the spec here)

**Comment by greg-man** [comment]
Ah! Already wondered why the tracks aren't removed directly after they are stopped instead relying on the event to be fired...

@fippo Any ideas how to fix this? I tried to call the stopScreenShare() method of the LocalMedia object instead of just stopping the track, but got caught in an infinite loop.

**Comment by tgabi333** [comment]
connect https://github.com/otalk/localmedia/pull/22

**Comment by tgabi333** [comment]
Fixed with 2.2.2 release
